### ISSUE TITLE
Doubt about semantics

### ISSUE THREAD

**Issue by teja5832** [p-org/P#72]
In the file Tst\RegressionTests\Feature2Stmts\Correct\receive12\receive12.p, line 20 says
b = new B(this);

However, B's start state's entry action Init does not take an argument - it looks like
machine B {
	start state Init {
		entry {
			raise Unit;
		}
                ...
This seems to pass the type check. What are the semantics of this? Is the argument 'this' dropped?

**Comment by shazqadeer** [comment]
Yes, the argument "this" is dropped.  When an action handler does not declare a payload variable, it is automatically generated with type ANY.  That is why, the type checker does not complain.  Perhaps the type of the auto-generated payload variable should be changed to the NULL type.  

@desaiankush: We had a discussion about this issue.  What conclusion did we come to?

**Comment by teja5832** [comment]
In that case, we can also do 
b = new B();
without any type error. How does that work?

**Comment by shazqadeer** [comment]
new B() is treated as new B(null).  The value null is actually passed into the entry function at runtime.  null is also a member of ANY type.

**Comment by shazqadeer** [comment]
@desaiankush :  Could you please remind me of the discussion we had about the issues @teja5832  is raising?

**Comment by teja5832** [comment]
Is that how all other anonymous functions are treated too?

**Comment by desaiankush** [comment]
Sorry about the late reply. I missed this thread.

@shazqadeer In the modp branch, we have made the payload type of an action handler as NULL if it does not declare payload argument.
@shazqadeer after this change the updated rule for constructor type is as follows:

	[ rule_Classes = '"error, msg: Payload type of constructor must be subtype of payload type of entry function of start state;"']
	PayloadError(n) :- SubSE(c, n), n = New(name, NIL), ConstructorType(m, _, ft), m.name = name, no MachineReceivesDecl(m, _), **ft != BaseType(NULL)**.

	[ rule_Classes = '"error, msg: Payload type of constructor must be subtype of payload type of the created interface;"']
	PayloadError(n) :- SubSE(c, n), n = New(name, NIL), InterfaceTypeDecl(name, _, ft), **ft != BaseType(NULL).**

@teja5832 : the updated type checking rule is that, you can do a new M() (where the payload type is NIL) only if the payload type of the constructor is NULL otherwise its a typing error.

**Comment by shazqadeer** [comment]
I have checked in a fix for the typing rules for new M(). There are two aspects to my change:
1. Type of default payload argument:  Payload argument is automatically generated in two situations.  First, an event handler is not supplied at all.  In this case, I have preserved the behavior and left the type of payload to be ANY.  Second, an event handler is supplied but is missing an explicit payload argument.  In this case, I have changed the behavior so the type of payload argument is NULL.
2. I have changed the typing rules for new M() as follows:
[ rule_Classes = '"error, msg: Entry function of start state expects an argument;"']
PayloadError(n) :- SubSE(c, n), n = New(name, NIL), ConstructorType(m, _, ft), m.name = name, ft != BaseType(ANY), ft != BaseType(NULL).

[ rule_Classes = '"error, msg: Entry function of start state expects an argument;"']
PayloadError(n) :- SubSE(c, n), n = NewStmt(name, NIL, _), ConstructorType(m, _, ft), m.name = name, ft != BaseType(ANY), ft != BaseType(NULL).

These rules requires if new M() is used then the type of payload argument of the entry function of the start state of M must be either ANY or NULL.  This means that new M() can always be used for any auto-generated payload type for the entry function of the start state because auto-generated type is either ANY or NULL.

After making this change, I had to fix many regression tests including the one report by @teja5832.  All those changes are also checked in.  I have not done the Zing regression tests.  The golden output of these might change a little bit.  Let us keep an eye on that.
